### PR TITLE
Module2

### PR DESCRIPTION
**Pull request by ptesavol**
# Add streamr-proto-rpc package

This is the c++ equvivalent of the typescript proto-rpc package of streamr-network.

## Differences compared to the typescript version

* outgoingMessageListener renamed to outgoingMessageCallback
* removed redundant emitting of outgoing messages as events
* generates its own client and server classes with its own plugin directly using google protoc
* split RpcCommunicator to 3 parts to increase code readability

### CODE REVIEW

**Comment by juslesan** [comment]
A general thought concerning the possibility of replacing the NodeJS version of this library with a wrapped version of this one. Would we need to generate the proto classes twice once for this C++ and once for TS

File: packages/streamr-proto-rpc/include/streamr-proto-rpc/RpcCommunicator.hpp [hunk 1]
```diff
@@ -0,0 +1,614 @@
...
+
+using google::protobuf::Any;
+using streamr::logger::SLogger;
+using RpcMessage = ::protorpc::RpcMessage;
+using RpcErrorType = ::protorpc::RpcErrorType;
+using folly::coro::Task;
+
+constexpr size_t defaultRpcRequestTimeout = 5000;
+
+/*
```
Comment (juslesan) [inline packages/streamr-proto-rpc/include/streamr-proto-rpc/RpcCommunicator.hpp]: Remove commented code blocks?

File: packages/streamr-proto-rpc/include/streamr-proto-rpc/RpcCommunicator.hpp [hunk 2]
```diff
@@ -0,0 +1,614 @@
...
+public:
+    explicit RpcCommunicator(
+        std::optional<RpcCommunicatorConfig> config = std::nullopt) { // NOLINT
+        objectId = RpcCommunicator::createObjectId();
+        if (config.has_value()) {
+            mRpcRequestTimeout = config.value().rpcRequestTimeout;
+        } else {
+            mRpcRequestTimeout = defaultRpcRequestTimeout;
+        }
+        mExecutor = new folly::CPUThreadPoolExecutor(29); // NOLINT
```
Comment (juslesan) [inline packages/streamr-proto-rpc/include/streamr-proto-rpc/RpcCommunicator.hpp]: Make the number of threads into a variable.

Comment (juslesan) [inline packages/streamr-proto-rpc/include/streamr-proto-rpc/RpcCommunicator.hpp]: Wondering if 29 threads is a bit much for an individual RpcCommunicator

File: packages/streamr-proto-rpc/include/streamr-proto-rpc/RpcCommunicator.hpp [hunk 3]
```diff
@@ -0,0 +1,614 @@
...
+    void rejectOngoingRequest(const RpcMessage& response) {
+        if (mStopped) {
+            return;
+        }
+        SLogger::trace("rejectOngoingRequest()", response.DebugString());
+        std::lock_guard lock(mOngoingRequestsMutex);
+
+        const auto& ongoingRequest = mOngoingRequests.at(response.requestid());
+
+        const auto& header = response.header();
```
Comment (juslesan) [inline packages/streamr-proto-rpc/include/streamr-proto-rpc/RpcCommunicator.hpp]: (Not relevant to this PR) We definitely should have made a proper proto definition for the headers before launch...

File: packages/streamr-proto-rpc/protos/ProtoRpc.proto [hunk 4]
```diff
@@ -0,0 +1,32 @@
+syntax = "proto3";
```
Comment (juslesan) [inline packages/streamr-proto-rpc/protos/ProtoRpc.proto:4]: A common place for the definitions would be very useful in the future.

File: packages/streamr-proto-rpc/include/streamr-proto-rpc/ServerRegistry.hpp [hunk 5]
```diff
@@ -0,0 +1,127 @@
...
+
+    template <typename T>
+    T getImplementation(
+        const RpcMessage& rpcMessage, const std::map<std::string, T>& map) {
+        if (!rpcMessage.header().contains("method")) {
+            throw UnknownRpcMethod(
+                "Header \"method\" missing from RPC message");
+        }
+
+        if (map.find(rpcMessage.header().at("method")) == map.end()) {
```
Comment (juslesan) [inline packages/streamr-proto-rpc/include/streamr-proto-rpc/ServerRegistry.hpp]: single variable `rpcMessage.header().at("method")`

File: packages/streamr-proto-rpc/include/streamr-proto-rpc/RpcCommunicator.hpp [hunk 6]
```diff
@@ -0,0 +1,614 @@
...
+    template <typename F>
+        requires std::is_assignable_v<OutgoingMessageCallbackType, F>
+    void setOutgoingMessageCallback(F&& callback) {
+        mOutgoingMessageCallback = std::forward<F>(callback);
+    }
+
+    // Client-side API
+    // This method is for making RPC calls only, not notifications
+    template <typename ReturnType, typename RequestType>
+    Task<ReturnType> callRemote(
```
Comment (juslesan) [inline packages/streamr-proto-rpc/include/streamr-proto-rpc/RpcCommunicator.hpp]: `callRemote` -> `request`
`doNotifyTask` -> `notify`

Comment (juslesan) [reply packages/streamr-proto-rpc/include/streamr-proto-rpc/RpcCommunicator.hpp]: Could follow similar naming convention everywhere. In NodeJS and here we have ie.`registerRpcMethod`. In the file methods are called methods, requests and calls. Could select one naming convention for clarity.

File: packages/streamr-proto-rpc/include/streamr-proto-rpc/RpcCommunicator.hpp [hunk 7]
```diff
@@ -0,0 +1,614 @@
...
+    void onIncomingMessage(
+        const RpcMessage& rpcMessage, const ProtoCallContext& callContext) {
+        SLogger::trace("onIncomingMessage", rpcMessage.DebugString());
+        std::lock_guard lock(mOngoingRequestsMutex);
+
+        const auto& header = rpcMessage.header();
+
+        SLogger::trace("onIncomingMessage() requestId", rpcMessage.requestid());
+        SLogger::trace("Printing all keys of mOngoingRequests:");
+        for (const auto& ongoingRequest : mOngoingRequests) {
```
Comment (juslesan) [inline packages/streamr-proto-rpc/include/streamr-proto-rpc/RpcCommunicator.hpp]: Remove this loop?

**Review by juslesan — COMMENTED** [review]
Some comments on RpcCommunicator and ServerRegistry

File: packages/streamr-proto-rpc/include/streamr-proto-rpc/RpcCommunicator.hpp (hunk 2 above)
Comment (ptesavol) [reply packages/streamr-proto-rpc/include/streamr-proto-rpc/RpcCommunicator.hpp]: done

File: packages/streamr-proto-rpc/protos/ProtoRpc.proto (hunk 4 above)
Comment (ptesavol) [reply packages/streamr-proto-rpc/protos/ProtoRpc.proto:4]: Any suggestions on this? I'll add a comment with pinned-down git reference to network-monorepo into these in aany case

File: packages/streamr-proto-rpc/include/streamr-proto-rpc/ServerRegistry.hpp (hunk 5 above)
Comment (ptesavol) [reply packages/streamr-proto-rpc/include/streamr-proto-rpc/ServerRegistry.hpp]: done

**Comment by ptesavol** [comment]
If using this component as a javascript native module, it would depend on the interface of the module if we would need to generate the protobuf classes also for typescript.

**Comment by ptesavol** [comment]
This pull request has been merged to pull request "Module3", and is not relevant anymore

**Pull request closed without merging**